katdal/datasources.py

```diff
@@ -168,37 +167,25 @@ def view_capture_stream(telstate, capture_block_id=None, stream_name=None):
     """
     # Detect the capture block
     if not capture_block_id:
-        capture_blocks = []
-        if 'sdp_capture_block_id' in telstate:
-            for value_time in telstate.get_range('sdp_capture_block_id', st=0):
-                cbid = str(value_time[0])
-                if 'obs_params' in telstate.view(cbid, exclusive=True):
-                    capture_blocks.append(cbid)
-        if not capture_blocks:
+        try:
+            capture_block_id = str(telstate['capture_block_id'])
+        except KeyError:
             raise ValueError('No capture block IDs found in telstate - '
                              'please specify it manually')
-        # Pick the latest capture block
-        capture_block_id = capture_blocks[-1]
-        if len(capture_blocks) > 1:
-            logger.warning('Telstate has more than one capture block - %s - '
-                           'picking the latest', capture_blocks)
     # Detect the captured stream
     if not stream_name:
-        streams = []
-        for stream, config in telstate['sdp_config']['outputs'].items():
-            if config['type'] != 'sdp.l0':
-                continue
-            capture_stream = telstate.SEPARATOR.join((capture_block_id, stream))
-            if 'chunk_info' in telstate.view(capture_stream, exclusive=True):
-                streams.append(stream)
-        if not streams:
+        try:
+            stream_name = str(telstate['stream_name'])
+            stream_type = str(telstate['stream_type'])
```

Comment:
Can we look up the stream type from within the `<stream>_` namespace? That's where it is now written by ingest, and it'll remove the need for the metadata writer to copy it to a top-level key.

Reply:
No - we are just rehashing the bikeshed now. There is no prior knowledge of the stream or cbid when katdal opens a file - we needed a simple, unambiguous way to tell what cbid/stream this file represents. The *only* way to ensure this is to have meta writer insert these values just before creating the file, and I don't want meta writer to actually mess with the live telstate.

Reply:
Sure, it needs to insert the name, but does it need to insert the type as well? Once you have the name you can get to the namespace to get the type (which is already in telstate).

Reply:
That is a good point. I'll just move the `.view(stream)` earlier, so similar effort for fewer telstate keys.